lib/jwt_sessions/authorization.rb

```diff
@@ -83,7 +83,10 @@ def retrieve_csrf
 
     def token_from_headers(token_type)
       raw_token = request_headers[JWTSessions.header_by(token_type)] || ''
-      token = raw_token.split(' ')[-1]
+      scheme, token = raw_token.split(' ')
```

Comment:
Hey @skrobul, thanks for the PR!
The issue is reasonable.
However, since jwt_sessions allows to configure a custom auth header name (‘Authorization’ on default), I think it should be possible also to configure a validation of HTTP scheme, as for some cases the validation might be uncessary. 
Thoughts?

Reply:
Good point, I didn't think of that. Given that providing HTTP authentication scheme makes sense mostly when the `Authorization` header is used, I think this could be solved in following ways:

1. Add new boolean configuration option (default to `true`) that can disable scheme validation
2. Verify if the scheme is `Bearer` **only** when access/refresh header is configured as `Authorization`. It probably doesn't make sense to do it for [other authentication schemes](https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml).
3. Combination of `1.` and `2.`.

Please  let me know what is your preference. I always try to use [standarized](https://tools.ietf.org/html/rfc7235#section-4.2) HTTP headers, so don't have much experience in using custom ones. However, I've seen `Authorization: JWT [token]` being used in the wild in which case `1.` would be needed.

Reply:
This will work, but I am not sure if these would be the best defaults. I have stumbled upon this problem upon receiving 'authorization errors' while the users were sending correct authentication credentials and valid cookie. It took a lot of  troubleshooting to identify cause of a problem.

If developer already knows what's causing the problem, they can simply work around this by changing `JWT.header_name` to something that will never occur  in real request, so I'm not sure if by providing `JWTSessions.http_scheme` in an opt-in model, we are adding any  value here.

At the same time, I understand you may want to keep the current behavior for backward compatibility. I was considering another alternative - would it be better if we just log a message stating that "Authorization header is configured, but HTTP scheme was not included in request" or something similar (i.e. custom exception that inherits from `Errors::Unauthorized` ?

Reply:
As understand your proposal, instead of a dedicated setting (http_scheme), we're going to have a new exception with a meaningful message for the cases when the http scheme is not provided in the request.
But there're several issues with this approach:
1. In your case the http scheme was provided in the request, but it was set to `Basic`. So this exact approach does not cover this particular case.
2. It's not backward compatible if we'll add a new exception for a case which was valid previously. 

Currently the gem always tries to auth by headers on default and in case of a failure it goes for a cookie based auth. I'm going to add more information to the documentation regarding the way it treats headers, http schemes and cookies. Also I though It'd be nice in addition to existing auth methods to provide dedicated ones, like `authorize_access_request_by_cookies!`/`authorize_access_request_by_headers!`.

Reply:
> Currently the gem always tries to auth by headers on default and in case of a failure it goes for a cookie based auth.

That's what I've seen in the README, but it's not really the case. jwt_sessions does not move on to cookie based authentication on failure. The `Errors::Unauthorized` is raised instead and no further checks are made.  Whole point of raising this  PR was to deal with this situation.  I guess it all boils down to what is considered a "__failure__".

 In my personal opinion, the default behaviour should be such that server attempts to verify credentials provided in a header and if they are invalid/unparseable/damaged, it should check if there are any credentials provided in a cookie. If there are, they should be checked before final decision is made.

> 1. In your case the http scheme was provided in the request, but it was set to Basic. So this exact approach does not cover this particular case.

Agreed. It would not solve the problem for end-users, but developer would know immediately what is causing a problem. This is just a proposal and might be a stupid idea :)

> 2. It's not backward compatible if we'll add a new exception for a case which was valid previously.

My point is that it wasn't 'valid'. Currently if request contains invalid Authorization header and valid credentials in a cookie, the `Errors::Unauthorized` is raised. If this behavior should remain unchanged, my proposal was to still raise an exception, just with slightly more descriptive message.

As developer using `jwt_sessions` I  found it bit not intuitive for users to get `Unauthorized` error even when they provide valid credentials, but they also happen to include extra header as well.

Reply:
I'm sorry, actually ignore the above. After reading the code bit more I realized that my `Errors::Unauthorized` might have been a result of invalid CSRF as well, so I'll try to reproduce the same code I have initially observed this problem in.

Reply:
> That's what I've seen in the README, but it's not really the case. jwt_sessions does not move on to cookie based authentication on failure. The Errors::Unauthorized is raised instead and no further checks are made. Whole point of raising this PR was to deal with this situation. I guess it all boils down to what is considered a "failure".

Sorry, by "failure" I meant the situation when the auth header is not found in the request headers.
If the header is present in the request but contains an invalid token then it won't fallback to a cookie based auth.